### ISSUE TITLE
ExpressionChangedAfterItHasBeenCheckedError when changing a component value in afterViewInit

### ISSUE THREAD

**Issue by adrienboulle** [angular/angular#14748]
**I'm submitting a ...**
```
[x] bug report
[ ] feature request
[ ] support request
```

**Current behavior**
changing a component value in afterViewInit throws `ExpressionChangedAfterItHasBeenCheckedError` if the value has an impact on a method used in the view.

in the plunker:
-> `getHeightPx()`  returns `0px` before  ngAfterViewInit is called (`isLoaded === undefined`) 
-> `getHeightPx()`  returns `100px` after  ngAfterViewInit is called (`isLoaded === true`) 

**Expected behavior**
it should not throw the error since `getHeightPx` is just a method and not part of the model.

**Minimal reproduction of the problem with instructions**
http://plnkr.co/edit/JdhPkjBl6wc0FIc8LcUs?p=preview

* **Angular version:** 4.0.0-rc.0 => KO
* **Angular version:** 4.0.0-beta.8 => OK

**Comment by DzmitryShylovich** [comment]
it's by design

**Comment by adrienboulle** [comment]
@DzmitryShylovich humm indeed just saw it, but then why did that work in 4.0.0-beta.8 ? (and many previous version since the release)

**Comment by DzmitryShylovich** [comment]
probably was a bug or it was disabled. anyway it throws in 2.x and it will throw in 4.x

**Comment by adrienboulle** [comment]
actually it does not
http://plnkr.co/edit/qjjtRtTcwAWflanmI7wr?p=preview

**Comment by DzmitryShylovich** [comment]
rc1 http://plnkr.co/edit/lU32vBJY9KYXZUTW7GQQ?p=preview

**Comment by DzmitryShylovich** [comment]
rc1 is the latest. 
https://github.com/angular/angular/blob/master/CHANGELOG.md :
> Note: the 4.0.0-rc.0 release on npm accidentally omitted one bug fix, so we cut rc.1 instead. oops :-)

**Comment by adrienboulle** [comment]
well I don't say it's not supposed to be like that, it's just that i've been updating model in afterViewInit since the release and never had this error

**Comment by DzmitryShylovich** [comment]
there's no error in rc1

**Comment by adrienboulle** [comment]
you have enableProdMode(); in your plunker, that's why

**Comment by DzmitryShylovich** [comment]
yeah, forgot to remote it. sorry.

**Comment by adrienboulle** [comment]
ok is see the difference.

- If we directly modify a value binded to the view it throws the error as expected (since 2.0.0 at least)
- If we modify a component value that indirectly modify the view (in the plunker `isViewInit` has an impact on the getTextDecoration method used in the view - via the ngStyle here) then it throws the error since 4.0.0-rc.1

http://plnkr.co/edit/I51JWge8P5oY0mj2Ni65?p=preview

**Comment by chunghha** [comment]
I also start to see ExpressionChangedAfterItHasBeenCheckedError after updating to 4.0.0-rc.1 from 4.0.0-beta.8 and 2.4.8.

https://plnkr.co/edit/atBNGvMgC3QxSGk7Sm6f?p=preview

Above plnkr shows Clarity's datagrid without a problem with 4.0.0-beta.8. But, after commenting out beta.8 and uncommenting rc1 in the config.js file, this error is thrown.

ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'Query 3 not dirty'. Current value: 'Query 3 dirty'.

Was there any breaking change regarding the above between beta.8 and rc.1? If I run it with enableProdMode(), the error disappears.

ping @youdz

**Comment by tbosch** [comment]
@adrienboulle I think you analysis is correct, and the error is intended in this case.

@chunghha this is a valid regression, as we didn't check queries for changes after checking before. Will create a PR to disable this again...

**Comment by tbosch** [comment]
Closing this in favor of https://github.com/angular/angular/issues/14925, as this tracks the query problem.

**Comment by chunghha** [comment]
can confirm my issue above fixed with rc.4. thanks!

**Comment by adrienboulle** [comment]
@tbosch It seems that the issue is still there:
http://plnkr.co/edit/ELk3mrdXIOcYIMxetjoV?p=preview

**Comment by DzmitryShylovich** [comment]
@adrienboulle open a new issue

**Comment by adrienboulle** [comment]
@DzmitryShylovich new issue here #15464

**Comment by minuz** [comment]
Hi all, 

If this error is by design, what's the correct pattern to avoid this sort of issue?
I have an observable and I'm using as follows 

```
<ul class="pagination pagination-flat pagination-xs no-margin-bottom" *ngIf="dataPages$ | async as dataPages">
            <li *ngFor="let item of dataPages" [ngClass]="{active: item.isActive, disabled: item.isDisabled}">
                <a *ngIf="item.isDisabled"><i class="{{item.icon}}" *ngIf="item.icon"></i>{{item.displayText}}</a>
                <a *ngIf="!item.isDisabled" (click)="onPageClicked(item.pageNum)">
                    <i class="{{item.icon}}" *ngIf="item.icon"></i>
                    {{item.displayText}}
                </a>
            </li>
        </ul>
```

The paging values will change once the data is loaded and the paging is calculated. It keep throwing that error on dev mode. In prod, that error does not appear. So I'm wondering what's the correct pattern to avoid this issue. 

Any help would be appreciated!

**Comment by jsJunky** [comment]
+1 on @minuz 's question.

I've built a carousel of sorts and the paging depends on the scrollWidth of child element. `@View` is always undefined during the initial check so I wait for ngAfterViewInit in order to set the paging. It appears to be the only life cycle event where I can reliably check the element exists.

Any other suggested solutions?

**Comment by aesopturtle** [comment]
On [this similar issue](https://github.com/angular/angular/issues/15464), it is suggested that we run the change detection explicitly after the change. It works for me.

https://stackoverflow.com/questions/39787038/how-to-manage-angular2-expression-has-changed-after-it-was-checked-exception-w

**Comment by jsJunky** [comment]
@aesopturtle Ended up with a similar solution. Had a slight issue where the change detection would fire and force children's afterInit fire which got me into a loop. But just added a boolean for the initialization and it fixed it up.

**Comment by minuz** [comment]
Hi All, I thought I should give a little update here.
I honestly haven't done anything and on my last update to angular 4.1.3 it stopped.
Did any of you noticed?

**Comment by CindySLiu** [comment]
I'm still getting the same error, even though I called this.changeRef.detectChanges(); at the end of ngAfterViewChecked or ngDoCheck.

Here is my error message:
container.component.html:4 ERROR Error: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'true'. Current value: 'false'.
    at viewDebugError (errors.ts:30)
    at expressionChangedAfterItHasBeenCheckedError (errors.ts:29)
    at checkBindingNoChanges (util.ts:145)
    at checkNoChangesNodeInline (view.ts:472)
    at checkNoChangesNode (view.ts:556)
    at debugCheckNoChangesNode (services.ts:389)
    at debugCheckRenderNodeFn (services.ts:332)
    at Object.eval [as updateRenderer] (container.component.html:4)
    at Object.debugUpdateRenderer [as updateRenderer] (services.ts:309)
    at checkNoChangesView (view.ts:371)
    at callViewAction (view.ts:735)

My container has
<nav class="navbar navbar-default navbar-fixed-top sb-slide" style="transform: translate(164px);">
    <ul class="navbar-icons">
        <li *ngFor="let item of actions" [ngClass]="item.cssClass" **[class.menuInactive]="!item.active"**>
            <a (click)="onActionClick(item.name)"> {{' ' + item.name + ' '}} </a>
        </li>
    </ul>
</nav>

In ngAfterViewChecked or ngDoCheck, I have the following to update menu active status and get the same error message every time when menu status is changed.
this.container.updateActions([
                    { name: "Copy", active: true },
                    { name: "Delete", active: true },
                ]);
this.changeRef.detectChanges();

I have Angular 4.1.3.

**Comment by CindySLiu** [comment]
What is the appropriate place to update UI menu to avoid this error?

**Comment by leocaseiro** [comment]
I've removed most of my warnings using `*ngIf` and `?`, however, I wonder if it's the best solution for this issue.

**Comment by minuz** [comment]
@CindySLiu I believe playing with the ngAfterViewChecked o ngDoCheck won't solve your problem as the nature of this error is having multiple changeDetection happening to the same object before finishing its rendering.

What you should try to do instead is preventing the change detection to be triggered.

Can you post a sample of your component where this issue is happening or a plunkr?

**Comment by zoechi** [comment]
You can inject `this.cdRef:ChangeDetectorRef` and then call `this.cdRef.detectChanges()` after you modified the model to work around the error.

**Comment by CindySLiu** [comment]
@minuz Here is the plunker url to reproduce this problem: https://plnkr.co/edit/jMfWDdisy6JrMhNb3yko

This is my first plunker app. I don't know why I got 404 error - 
zone.js:2263 GET https://run.plnkr.co/trQwvmYSIBehpeRB/container.component.html 404 ().

If you download the zip, it works fine. All the files are there. I don't know why plunker cannot load the files I created.

In my app, I have a container that contains dynamically loaded components. The component triggers menu status change in the container. In the component's ngAfterViewChecked, I have this.changeRef.detectChanges();

This is not enough. I have to put this.changeRef.detectChanges(); in the container's ngAfterViewChecked to prevent the error (I just found out).

**Comment by minuz** [comment]
Hi @CindySLiu, your plunkr seems incomplete. There's nothing running there. The nav bar does not have any items and the content is not loaded. When opening the console, I also cannot see the error, apart from the 404.

Maybe create a git repo with a working version. I guess it will be easier to help you out.

**Comment by TheBolmanator** [comment]
TL;DR  I will be ignoring this alarm and intend to do nothing to prevent seeing the red in the dev tools console, and leave it to 'others' to figure out what program failure this error is intended to flag and what the correct mitigation is when the error is thrown.  
__________________________________________________________________...
I am wondering if someone can help me understand...  why this is an error.  I mean - 'so things change after the view is initialized?  So what?'  Until now I (and my app) have been working happily without knowing that this 'error' existed and now that this error is being thrown I am supposed to change something, but no one seems quite sure what this change entails.  

I am approaching this subject from a background in process engineering where there has been a lot of research and learning on the subject of ["Alarm Management"](https://en.wikipedia.org/wiki/Alarm_management) and it seems to me that there is an application of these learnings that may be made in this situation.   

As I see it there is sound reason for pointing out to developers that there is a behavior that is outside the optimal case,  HOWEVER I do not agree that this constitutes an error.  If the resultant system behavior has not changed (meaning there is an error thrown that was not previously/consistently thrown), but the system that generates errors (the alarm system) has been modified... (here I will revert to "Alarm Management" terminology)  and there is an alarm, for which there is no clearly defined operator response, and there is no clearly defined escalation hazard.  The only outcome is to train the operators to ignore the alarm.  This results in a reduction in total safety.  

This is why, in the process industries where failure to properly respond to 'real' hazards, can result in catastrophic outcomes; a great deal of time is spent on alarm rationalization and the bias is on removal of alarms not generating them. 

I mean no offence, and trust that some allowance may be made for my ignorance of the nuance of this issue, and or what may or may not be appropriate in GitHub.

**Comment by zoechi** [comment]
> The only outcome is to train the operators to ignore the alarm.

That's not  true. It points out a possible error or architectural issue and it easy to fix
There is no need to ignore it.

**Comment by intellix** [comment]
Hope it helps someone who stumbles upon this issue through Google. I get this quite frequently and I still don't have a clue how I'm supposed to be fixing it properly tbh.

I've got a chat in which I use a currentUser$ observable to display admin functionality so I do something like:

```ts
currentUser$ = this.userService.currentUser$;
```

```html
<xc-message-list [user]="user | async" [messages]="messages$ | async"></xc-message-list>
```

When you click `Logout` in the header then the current user becomes null and so my views need to update. I don't understand what I'm doing wrong here but the hack/workaround/fix that worked for me:

```ts
currentUser$ = this.userService.currentUser$.delay(1);
```

Would love to be enlightened about how my code is architecturally wrong.

**Comment by victornoel** [comment]
@zoechi clearly what @TheBolmanator is saying is that no, there is no easy fix!

In this thread we see a lot of unanswered questions about how to manage this… This one is a good example: https://github.com/angular/angular/issues/14748#issuecomment-300976768.
It seems that some of the error disappear if you play with ngIf and as in order to change the way things are interpreted at runtime… I'm pretty sure this is neither an easy fix (hard to maintain) and not an architectural issue…

It is not clear what is the source of these errors.
Yes, in some cases, the problem is clear and you need to force change detection, but for the rest, there is many cases where you have an async pipe that makes this problem appear with no reason whatsoever.

**Comment by zoechi** [comment]
@victornoel comments without a Plunker that demonstrates the problem are not really helpful

**Comment by TheBolmanator** [comment]
TL;DR  I will say no more on github, so please feel free to ignore my post above, it was a mistake.

@zoechi,  Please refrain from making an unsupported claim that what I say is "not true"  (I am a Texan and old enough to still consider "them's fighting words!").  I will choose, however, to interpret your statement as rather an awkward way of saying that you do not agree.  Rest assured your position on this may be inferred from your various posts on the subject elsewhere.    

As an experiment, I implemented the "workaround" that you recommend, calling this.cdRef.detectChanges(), and when placed correctly this did remove the ExpressionChanged... error alert,  however the result was to reintroduced several problems in the rendered page that had been addressed (albeit without elegance) via application of setTimeout().  I would add that the elegance of your workaround is lost on me.

I had two reasons for my previous post - A.) to share that _in my case_ ignoring the ExpressionChanged... error messages had proved to be a completely rational and acceptable solution.  B.) to share my concern that throwing "errors" on code that both compiles and renders correctly, because this _might_ not align with someone's idea of "Correct" architectural pattern, is an antipattern.

**Comment by zoechi** [comment]
@TheBolmanator please refrain from making claims that are based on your limited Angular knowledge ;-)
When you complain about Angular without exact instructions how to reproduce the problem, it's IMHO safe to treat the comments as spam.

**Comment by jsJunky** [comment]
🍿

**Comment by leocaseiro** [comment]
I found 3 solutions for who are looking for easy fixes.

1. I've moved from [ngAfterViewInit](https://angular.io/api/core/AfterViewInit) to [ngAfterContentInit](https://angular.io/api/core/AfterContentInit)

2. I've used [ChangeDetectorRef](https://stackoverflow.com/a/39787056/3415716) as suggested  on https://github.com/angular/angular/issues/14748#issuecomment-307291715

3. Moving to [ngAfterViewChecked](https://angular.io/api/core/AfterViewChecked) also fixes.

**Comment by victornoel** [comment]
And I have another strange one, I get the `ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: '[object Object]'. Current value: ''.` and when I check with the debugger, the previous value is `{ wrapped: '' }`.

Where does this wrapped comes from? I feel like the value does not actually change, but the way angular initialise things (I have an `Observable` from the ngrx store and I call `(myObs$ | async).properties` on it) makes a simple `async` use fail!

I have this problem in multiple places with other type of values. I'm trying to make a plunkr but I'm struggling a bit for now :)

**Comment by tytskyi** [comment]
@victornoel is your problem related to https://github.com/angular/angular/issues/15721?

**Comment by victornoel** [comment]
@tytskyi yes, I was actually going to add a comment :)

**Comment by pappacurds** [comment]
was able to work around this be using ngDoCheck() method and calling the ChangeDetectorRef.detectChanges method. 

    public ngDoCheck(): void {
        this.cdr.detectChanges();
    }

**Comment by tomonari-t** [comment]
Hi this is a helpful post to understand and fix this error.

[Everything you need to know about the `ExpressionChangedAfterItHasBeenCheckedError` error](https://hackernoon.com/everything-you-need-to-know-about-the-expressionchangedafterithasbeencheckederror-error-e3fd9ce7dbb4)

**Comment by imransilvake** [comment]
this worked for me.

`public ngDoCheck(): void {
    this.cdr.detectChanges();
}`

**Comment by victornoel** [comment]
@tytskyi actually, after some research, my problem is this one #18129.

**Comment by Markus-ipse** [comment]
@pappacurds and @imransilvake 
That's not a great solution, what your doing is telling angular to run change detection every time change detection has occured, in other words, you've created an infinite loop..

**Comment by zoechi** [comment]
Just call `detectChanges()` only immediately after changing the value, then you don't get an infinite loop.

**Comment by MikeDub** [comment]
I had same issue when upgrading to angular 4.2.x.
Found the issue was in an EventSubscription, adding this.cdr.detectChanges(); straight after the value change in the subscription setup helped.

**Comment by dgroh** [comment]
Having the same issue here with nested router-outlets when using subscription:

https://github.com/angular/angular/issues/15634

**Comment by minuz** [comment]
Hi All, 

I was having this issue and it was bugging me for a long time. I think I finally found a pattern that is avoiding receiving this problem. It seems to be essentially a synchronization issue.

Approach one: use NGRX and start managing the state of the objects you are handling and getting this sort of issue. To implement this, find the many tutorials available on the internet. Imo, this approach is only good when you are managing state of many objects across many components. Otherwise it adds a development overhead that you need to evaluate if truly is worthwhile.

Approach two: create a simple service that manages the state of this object using BehaviorSubject.

For this approach, check this out.

```
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import { Injectable } from '@angular/core';

@Injectable()
export class SecurityLayoutStateService {

    addNewbuttonState$: BehaviorSubject<boolean>;

    constructor() {
        this.addNewbuttonState$ = new BehaviorSubject(false);
    }

    getAddNewButtonState() {
        return this.addNewbuttonState$.asObservable;
    }

    setAddNewButtonState(value: boolean) {
        this.addNewbuttonState$.next(value);
    }
}
```

So lets say you have a root component that is responsible to set the basic layout of the component and multiple sub-components that are basically different content but can re-use the root structure. The button Add new should only appear on the listing content and hide on edit mode.

```
<page-header [pageTitle]="'Security Management'"></page-header>
<div class="panel panel-white">
    <div class="panel-heading">
        <h6 class="panel-title">{{title}}</h6>
        <div class="heading-elements" *ngIf="showButton">
            <a class="btn btn-default" (click)="btnAddNew.emit()">
                <i class="icon-add position-left"></i> Add new
            </a>
        </div>
    </div>
    <div class="dataTables_wrapper no-footer">
        <div class="tabbable">
            ... omitted for brevity
           <router-outlet></router-outlet>
        </div>
        <div class="datable-footer panel-footer" *ngIf="isPagingAvailable">
            <span class="heading-text text-muted p-15">Showing page {{paging.currentPage$ | async}} out of {{paging.totalPages$ | async}}. Total items: {{paging.totalItems$ | async}}</span>
            <paging-component class="pull-right" (onLoad)="onLoad.emit($event)" #paging></paging-component>
        </div>
    </div>
</div>
```

The content received inside the `<router-outlet></router-outlet>` modifies the value of the `showButton` and throwing the annoying exception we are all seeing.

Modified the showButton to the following:
```
@Component({
    moduleId: module.id,
    selector: 'app-security',
    templateUrl: 'security.root.component.html'
})
export class SecurityRootComponent implements OnInit {
    constructor(private layoutStateService: SecurityLayoutStateService) { }

    ngOnInit() {
        ....
        this.addNewbuttonState$ = this.layoutStateService.getAddNewButtonState();
    }
}
```
`<div class="heading-elements" *ngIf="addNewbuttonState$ | async">...</div>`

Now that annoying message is no longer showing up and the load stays correct.
Whenever you need to change the state of the button, simply inject the service and use the setter. Whatever component is subscribed to this Observable will get the update.

I hope this can help you all.
Cheers

**Comment by realappie** [comment]
@victornoel The wrapped value comes from [this](https://v2.angular.io/docs/ts/latest/api/core/index/WrappedValue-class.html) class. It has something to do with pipes.

**Comment by sandraacha9** [comment]
Gracias por sus comentarios. Ahora tengo mas claro el problema, su causa y solución.
Este articulo me ha servido de mucha ayuda: https://blog.angularindepth.com/everything-you-need-to-know-about-the-expressionchangedafterithasbeencheckederror-error-e3fd9ce7dbb4

**Comment by enoh-barbu** [comment]
@sandraacha9 nu inteleg ce spui...

**Comment by tisak** [comment]
you can find it nice implementation of detectChanges() fix here:
https://stackoverflow.com/questions/48457666/expressionchangedafterithasbeencheckederror-in-ngoninit
